aiida_quantumespresso/workflows/pdos.py

```diff
@@ -0,0 +1,564 @@
+
+
+def validate_dos_parameters(value, _):
+    """Validate DOS parameters.
+
+    - shared: Emin | Emax | DeltaE
+    - dos.x only: ngauss | degauss | bz_sum
+    - projwfc.x only: ngauss | degauss | pawproj | n_proj_boxes | irmin(3,n_proj_boxes) | irmax(3,n_proj_boxes)
+
+    """
+    jsonschema.validate(value.get_dict(), get_parameter_schema())
+
+
+def clean_calcjob_remote(node):
+    """Clean the remote directory of a ``CalcJobNode``."""
+    cleaned = False
+    try:
+        node.outputs.remote_folder._clean()  # pylint: disable=protected-access
+        cleaned = True
+    except (IOError, OSError, KeyError):
+        pass
+    return cleaned
+
+
+def clean_workchain_calcs(workchain):
+    """Clean all remote directories of a workchain's descendant calculations."""
+    cleaned_calcs = []
+
+    for called_descendant in workchain.called_descendants:
+        if isinstance(called_descendant, orm.CalcJobNode):
+            if clean_calcjob_remote(called_descendant):
+                cleaned_calcs.append(called_descendant.pk)
+
+    return cleaned_calcs
+
+
+PwBaseWorkChain = plugins.WorkflowFactory('quantumespresso.pw.base')
+DosCalculation = plugins.CalculationFactory('quantumespresso.dos')
+ProjwfcCalculation = plugins.CalculationFactory('quantumespresso.projwfc')
+
+
+class PdosWorkChain(ProtocolMixin, WorkChain):
+    """A WorkChain to compute Total & Partial Density of States of a structure, using Quantum Espresso."""
+
+    @classmethod
+    def define(cls, spec):
+        # type: (ProcessSpec) -> None
+        # yapf: disable
+        """Define the process specification."""
+        super().define(spec)
+        spec.input('structure')
+
+        spec.expose_inputs(
+            PwBaseWorkChain,
+            namespace='base_scf',
+            exclude=('clean_workdir', 'pw.structure', 'pw.parent_folder'),
+            namespace_options={
+                'help': 'Inputs for the `PwBaseWorkChain` to run `scf` and `nscf` calculations.',
+                'validator': validate_base_scf,
+                'required': False,
+                'populate_defaults': False,
+            }
+        )
+        spec.expose_inputs(
+            PwBaseWorkChain,
+            namespace='base_nscf',
+            exclude=('clean_workdir', 'pw.structure'),
+            namespace_options={
+                'help': 'Optional inputs for `nscf` calculation, that override those set in the `base` namespace.',
+                'validator': validate_base_nscf
+            }
+        )
+        spec.input(
+            'parameters',
+            valid_type=orm.Dict,
+            serializer=to_aiida_type,
+            validator=validate_dos_parameters,
+            help='Combined input parameters for the `dos.x` and `projwfc.x` calculations.'
+        )
+        spec.expose_inputs(DosCalculation, namespace='dos', exclude=('parent_folder', 'parameters'))
+        spec.expose_inputs(ProjwfcCalculation, namespace='projwfc', exclude=('parent_folder', 'parameters'))
```

Comment:
I think this is because these parameters _need_ to be the same for these two calculations? @chrisjsewell correct me if I'm wrong. We could also add a validator to check this instead though, that would indeed be more consistent. There are also the `dos_only` and `proj_only` keys in the `parameters`, so clearly not all of them need to be the same.

Reply:
Have adjust this, and made a separate input for the `align_to_fermi` input. I've also made some changes here, since I think setting this to True only makes sense in case the `Emin` and `Emax` are actually provided and not extracted from the `nscf` calculation. I've also added some validation for this point, as well as making sure they are the same for the `dos` and `projwfc` calculations.